README.md
# Miftah-KeysGenerator
Miftah is a license keys generator API, comes whith a React Native mobile app.

## Getting Started

These instructions will get you a copy of the project up and running on your mobile phone or by direct using the API. 

### Installing

How to get and install the app 

Download the app.apk from ????

Install the app

End with an example of getting some data out of the system or using it for a little demo

## Using the APP

After opening the app you have to fill the fields and click the Generate button. This will get you a list of your keys sorted by date.

## Built With

* [ReactNative] (https://reactnative.dev/) - The mobile framework used
* [NodeJs] (https://nodejs.org/en/) - The web framework used


## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available.

## Authors

*-** CUB1 ** - *Initial work* - [Miftah-KeysGenerator](https://github.com/youcefislam/KeyGenerator)


## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details

## Acknowledgments

* Hat tip to anyone whose code was used


changelog.md

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2020-03-02
### First Added
--RESTFUL API for the generator app.
- Start by defining a generator function to generate a random keys.
- Create a Post Route to generate keys for specefic name.
- Create Mysql database to store the generated keys.
- Define Mysql queries for inserting and deleting.
--React Native mobile app .
- Create a simple input for the name.
- Create fetch request to send a generate request to the server.
- Add a button "Generate".
- Create list to show the generated keys.
- Add fetch  to send a delete request to the server.
- Add delete button


## [2.0.0] - 2020-03-03
### Added
- Add styling to the app.


## [3.0.0] - 2020-03-04
### Added
- Add checkbox for the upperCase choise.
- Add fields for the arguments(charts, groupes ,number).
- Add copy button to copy a key to the clipboard.

### Changed
- Add charts, groupes ,number to the generator function .
- Update the Routes.
- Improve argument against commit logs.
- Improve the queries to avoid the sql injection.
- Improve the app speed.
- Minimize the app size.



KeyServer/index.js
var express = require("express"),
    mysql = require('mysql'),
    bodyParser = require("body-parser");

//Create express app
app = express();
app.use(bodyParser.urlencoded({ extended: true }));
app.use(express.static("public"));
app.use(express.json({ limit: "1mb" }));
app.set('json spaces', 40);

//create mySql connection                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       
connection = mysql.createConnection({
    host: 'localhost',
    user: 'admin',
    password: 'password',
    database: 'mydb'                                                
});
connection.connect();                                                                       



//key Generator function
const generate = (charts, group, upper) => {
    const num = "1234567890";
    const upperCase = "ABCDEFGHIJKLMNOPQRSVUSWXYZ";
    const lowerCase = "abcdefghijklmnopqrsvuswxyz";

    if (upper == true) character = num + upperCase;
    else character = num + lowerCase;

    var myRandom = 0,
        myTextContent = "";
    for (i = 0; i < group * charts; i++) {
        myRandom = Math.floor(Math.random() * character.length);
        myTextContent += character[myRandom];
        if (((i + 1) % charts == 0) && (i != 0) && (i + 1 != group * charts)) {
            myTextContent += "-"
        }
    }
    return (myTextContent);
}

// generating the requested keys and put them into an array  
const genKeys = (charts, groupes, number, id_user, upper) => {

    var arr = [];
    //get today's date
    const date = new Date();
    //YYYY-MM-DD format
    const mysqlDate = date.toISOString().split("T")[0];

    for (let i = 0; i < number; i++) {
        var serial = generate(charts, groupes, upper);    // go to(generate function)
        arr.push([
            id_user,
            serial,
            mysqlDate
        ]);
    }
    return arr;
}


// Build response
const buildResponse = (keysArray, id) => {
    {
        let array = [];
        for (let i = 0; i < keysArray.length; i++) {
            array.push({
                id_key: id + i,  //add the Key id to the JSON array
                id_user: keysArray[i][0],
                serial: keysArray[i][1],
                key_date: keysArray[i][2]
            });
        }
        return array;
    };
}

//Post request , render a json array of generated keys
app.post("/key/create", async (req, res) => {

    if (req.body.name === '' || req.body.charts == '' || req.body.groupes == '' || req.body.number == '')
        res.json({ created: false });
    else {
        var name = req.body.name,
            charts = parseInt(req.body.charts),  //number of character by groupe { [x] [x] [x] [x] }-{ [x] [x] [x] [x] }-{ [x] [x] [x] [x] }
            groupes = parseInt(req.body.groupes), //number of groupes {xxxx}-{xxxx}-{xxxx}-{xxxx}
            number = parseInt(req.body.number),  //number of keys
            upper = req.body.upper;  //uppercase's code = true or false 

        var id_user;
        await connection.query("select * from usersTable where name = ?", name, (err, result) => { // testing if the name exist before
            if (err) throw err;
            id_user = result[0];
            if (id_user === undefined) {   //if the user do not exist

                connection.query("insert into usersTable(name) values('" + name + "')", (err, result) => {
                    if (err) throw err;

                    //Save the keys to the database
                    var keysArray = genKeys(charts, groupes, number, result.insertId, upper); //got to genkeys function 
                    const query = "insert into keysTable(id_user, serial, key_date) values ?";

                    connection.query(query, [keysArray], (err, result) => { // insert the keys array to the database
                        if (err) throw err;
                        var array = buildResponse(keysArray, result.insertId);  // return the keys list                     
                        res.json(array);    // send the list of keys within the response 
                    });
                });
            } else {
                //Save the keys to the database
                var keysArray = genKeys(charts, groupes, number, id_user.id_user, upper);
                const query = "insert into keysTable(id_user, serial, key_date) values ?";

                connection.query(query, [keysArray], (err, result) => {
                    if (err) throw err;
                    var array = buildResponse(keysArray, result.insertId);   // return the keys list
                    res.json(array);    // send the list of keys within the response 
                });
            }
        });
    }
});



// Delete request , render a json response of generated keys
app.post("/key/delete", (req, res)=>{
	const id_key = req.body.id_key; 
	const query = "delete from keysTable where id_key = ?";
	connection.query(query, id_key, (err)=>{            // delete query 
		if(err) {                       
			res.json({
				deleted: false                  
			});
		}
		else{
			res.json({
				deleted: true
			});
		}
	});
});



app.listen(3000, function () {      
    console.log("server started");
});



Miftah/Miftah/Screens/KeyList.js
import 'react-native-gesture-handler';
import React, { useState } from 'react'
import {
    Text,
    View,
    StyleSheet,
    TouchableOpacity,
    SafeAreaView,
    FlatList,
    Clipboard,
    ToastAndroid
} from 'react-native'


const KeysList = ({ route }) => {

    const [Keys, setKeys] = useState(route.params);   // List of The keys

    // Render Items Function (View for each Key)
    const ListItem = ({ item }) => {
        return (
            <View style={styles.item}>
                <TouchableOpacity style={styles.copyButton} onPress={() => CopyItem(item.serial)}>
                    <Text style={styles.buttonText}>Copy</Text>
                </TouchableOpacity>
                <Text style={styles.serialText}>{item.serial}</Text>
                <TouchableOpacity style={styles.deleteButton} onPress={() => DeleteItem(item.id_key)}>
                    <Text style={styles.buttonText}>Delete</Text>
                </TouchableOpacity>
            </View>
        );

    };

    // Copy to clipboard Function
    const CopyItem = (serial) => {
        Clipboard.setString(serial);
        ToastAndroid.show('Key Copied !', ToastAndroid.SHORT);
    }
    //end of Clipboard Function

    //Delete Keys Function (by ID)
    const DeleteItem = (id_key) => {

        const data = { id_key: id_key };
        
        fetch('http://aymengh-fjnuc.run-us-west2.goorm.io/key/delete', {   //make a post request for the key generator api
            method: 'POST',
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(data),    
        })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error('Something went wrong on our api server'); 
                }                
            })
            .then((responseJson) => {
                if (responseJson.deleted == true) {      // verify if the request has done to delete it from our flat list
                    setKeys((prevKeys) => {
                        return prevKeys.filter(item => item.id_key != id_key)
                    });
                    ToastAndroid.show('Key Deleted !', ToastAndroid.SHORT);
                }else{
                    alert("Something went wrong on our api server !"); 
                }
            })
            .catch(error => {
                console.error(error);
            });

    }
    // End of Delete Keys Function


    return (
        <SafeAreaView style={styles.container}>
            <FlatList
                data={Keys}                                             // flat List Interface
                renderItem={({ item }) => ListItem({ item })}   // render items using ListItem function GoTo(ListItem)
                keyExtractor={(item) => item.id_key}           
            /> 
        </SafeAreaView>
    );

}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#FA6148',
    },


    item: {
        flexDirection: 'row',
        backgroundColor: '#FDFCFB',
        padding: 10,
        marginVertical: 4,
        marginHorizontal: 10,
        borderRadius: 10,
    },


    deleteButton: {
        width: '17%',
        marginLeft: 2,
        alignItems: 'center',
        justifyContent: 'center',
        borderStyle: 'solid',
        borderColor: '#FA6148',
        borderWidth: 1,
        borderRadius: 10
    },
    buttonText: {
        color: '#c71f16',
    },
    sortByButton: {
        alignItems: "center",
        height: 23, width: '97%',
        backgroundColor: "#FA6148",
        borderWidth: 1,
        borderColor: 'white',
        borderBottomStartRadius: 5,
        borderBottomEndRadius: 5
    },
    sortBytext: {
        color: 'white',
        fontSize: 15
    },
    copyButton: {
        justifyContent: 'center',
    },
    serialText: {
        flex: 4,
        fontSize: 15,
        borderLeftWidth: 1,
        paddingLeft: 5,
        borderLeftColor: '#c71f16',
        color: '#242424',
        marginLeft: 5
    },
});

export default KeysList;

Miftah/Miftah/ios/Miftah/AppDelegate.m
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "AppDelegate.h"

#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge
                                                   moduleName:@"Miftah"
                                            initialProperties:nil];

  rootView.backgroundColor = [[UIColor alloc] initWithRed:1.0f green:1.0f blue:1.0f alpha:1];

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  return YES;
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

@end


Miftah/Miftah/ios/MiftahTests/MiftahTests.m
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <UIKit/UIKit.h>
#import <XCTest/XCTest.h>

#import <React/RCTLog.h>
#import <React/RCTRootView.h>

#define TIMEOUT_SECONDS 600
#define TEXT_TO_LOOK_FOR @"Welcome to React"

@interface MiftahTests : XCTestCase

@end

@implementation MiftahTests

- (BOOL)findSubviewInView:(UIView *)view matching:(BOOL(^)(UIView *view))test
{
  if (test(view)) {
    return YES;
  }
  for (UIView *subview in [view subviews]) {
    if ([self findSubviewInView:subview matching:test]) {
      return YES;
    }
  }
  return NO;
}

- (void)testRendersWelcomeScreen
{
  UIViewController *vc = [[[RCTSharedApplication() delegate] window] rootViewController];
  NSDate *date = [NSDate dateWithTimeIntervalSinceNow:TIMEOUT_SECONDS];
  BOOL foundElement = NO;

  __block NSString *redboxError = nil;
#ifdef DEBUG
  RCTSetLogFunction(^(RCTLogLevel level, RCTLogSource source, NSString *fileName, NSNumber *lineNumber, NSString *message) {
    if (level >= RCTLogLevelError) {
      redboxError = message;
    }
  });
#endif

  while ([date timeIntervalSinceNow] > 0 && !foundElement && !redboxError) {
    [[NSRunLoop mainRunLoop] runMode:NSDefaultRunLoopMode beforeDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
    [[NSRunLoop mainRunLoop] runMode:NSRunLoopCommonModes beforeDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];

    foundElement = [self findSubviewInView:vc.view matching:^BOOL(UIView *view) {
      if ([view.accessibilityLabel isEqualToString:TEXT_TO_LOOK_FOR]) {
        return YES;
      }
      return NO;
    }];
  }
  
#ifdef DEBUG
  RCTSetLogFunction(RCTDefaultLogFunction);
#endif

  XCTAssertNil(redboxError, @"RedBox error: %@", redboxError);
  XCTAssertTrue(foundElement, @"Couldn't find element with text '%@' in %d seconds", TEXT_TO_LOOK_FOR, TIMEOUT_SECONDS);
}


@end


Miftah/Miftah/android/app/src/main/java/com/miftah/MainApplication.java
package com.miftah;

import android.app.Application;
import android.content.Context;
import com.facebook.react.PackageList;
import com.facebook.react.ReactApplication;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.ReactPackage;
import com.facebook.soloader.SoLoader;
import java.lang.reflect.InvocationTargetException;
import java.util.List;

public class MainApplication extends Application implements ReactApplication {

  private final ReactNativeHost mReactNativeHost =
      new ReactNativeHost(this) {
        @Override
        public boolean getUseDeveloperSupport() {
          return BuildConfig.DEBUG;
        }

        @Override
        protected List<ReactPackage> getPackages() {
          @SuppressWarnings("UnnecessaryLocalVariable")
          List<ReactPackage> packages = new PackageList(this).getPackages();
          // Packages that cannot be autolinked yet can be added manually here, for example:
          // packages.add(new MyReactNativePackage());
          return packages;
        }

        @Override
        protected String getJSMainModuleName() {
          return "index";
        }
      };

  @Override
  public ReactNativeHost getReactNativeHost() {
    return mReactNativeHost;
  }

  @Override
  public void onCreate() {
    super.onCreate();
    SoLoader.init(this, /* native exopackage */ false);
    initializeFlipper(this); // Remove this line if you don't want Flipper enabled
  }

  /**
   * Loads Flipper in React Native templates.
   *
   * @param context
   */
  private static void initializeFlipper(Context context) {
    if (BuildConfig.DEBUG) {
      try {
        /*
         We use reflection here to pick up the class that initializes Flipper,
        since Flipper library is not available in release mode
        */
        Class<?> aClass = Class.forName("com.facebook.flipper.ReactNativeFlipper");
        aClass.getMethod("initializeFlipper", Context.class).invoke(null, context);
      } catch (ClassNotFoundException e) {
        e.printStackTrace();
      } catch (NoSuchMethodException e) {
        e.printStackTrace();
      } catch (IllegalAccessException e) {
        e.printStackTrace();
      } catch (InvocationTargetException e) {
        e.printStackTrace();
      }
    }
  }
}


Miftah/Miftah/Screens/Home.js
import 'react-native-gesture-handler';
import React, { useState } from 'react';
import CheckBox from '@react-native-community/checkbox';
import {
  StyleSheet,
  ActivityIndicator,
  Text,
  TextInput,
  TouchableOpacity,
  KeyboardAvoidingView,
  View,
  StatusBar,

} from 'react-native';


const Home = ({ navigation }) => {

  const [name, setName] = useState("");
  const [charts, setCharts] = useState("");
  const [groupes, setGroupe] = useState("");
  const [number, setNumber] = useState("");
  const [stats, setStats] = useState(false);
  const [upCase, setupCase] = useState(false)

  // Generating key and Navigate into the keys List
  const GoToKeyList = () => {
    const data = {
      name: name,
      charts: charts,
      groupes: groupes,
      number: number,
      upper: upCase
    };

    // requesting key from the server
    fetch('http://aymengh-fjnuc.run-us-west2.goorm.io/key/create', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    })
      .then(response => {
        if (response.status === 200) {
          return response.json();
        } else {
          throw new Error('Something went wrong on our api server');
        }
      })
      .then((responseJson) => {
        if (responseJson.created == false) {
          alert(' all fields must be filled !')
        }
        else {
          navigation.navigate('Key List', responseJson); //navigate to keys list screen with the response(keys list)
        }
        setStats(false);
      })
      .catch(error => {
        console.error(error);
        setStats(false);
      });
  }

  return (
    <KeyboardAvoidingView style={styles.container} contentContainerStyle={styles.afterFocus} behavior="position" >
      <StatusBar backgroundColor="#FA6148" barStyle="light-content" />
      <View style={styles.upperCasestyle}>
        <Text>UpperCase</Text>
        <CheckBox
          value={upCase}
          onChange={() => { upCase ? setupCase(false) : setupCase(true) }} />
      </View>
      <TextInput
        style={styles.inpuField}    
        placeholder="Name"
        onChangeText={(value) => setName(value)}      //take input from the user and put it into name variable
        autoCapitalize='none'
        placeholderTextColor='#FA6148'   />
      <TextInput
        style={styles.inpuField}
        placeholder="Groupe (<10)"
        onChangeText={(value) => setGroupe(value)}
        keyboardType='numeric'
        maxLength={1}
        placeholderTextColor='#FA6148' />
      <TextInput
        style={styles.inpuField}
        placeholder="Charts (<10 )"
        onChangeText={(value) => setCharts(value)}
        keyboardType='numeric'
        maxLength={1}
        placeholderTextColor='#FA6148' />
      <TextInput
        style={styles.inpuField}
        placeholder="Keys to Generate (<100)"
        onChangeText={(value) => setNumber(value)}
        keyboardType='numeric'
        maxLength={2}
        placeholderTextColor='#FA6148' />
      <TouchableOpacity style={styles.generateButton} onPress={() => { GoToKeyList(); setStats(true)}} >
        <Text style={styles.generateButtonText}>Generate</Text>  
      </TouchableOpacity>
      {
        stats ? <ActivityIndicator size="large" color="#FA6148" /> : <View></View>
      }
    </KeyboardAvoidingView >
  );
};



const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white',
    alignItems: 'center',
    justifyContent: "space-around",
  },
  afterFocus: {
    height: '80%',
    top: 1,
  },

  generateButton: {
    backgroundColor: 'white',
    marginTop: 50,
    marginBottom: 10,
    borderStyle: 'solid',
    borderWidth: 2,
    borderColor: '#FA6148',
    borderRadius: 20,
    alignItems: 'center',
  },
  generateButtonText: {
    fontSize: 20,
    color: '#FA6148',
  },

  inpuField: {
    marginBottom: 20,
    fontSize: 15,
    backgroundColor: "#FDFCFB",
    borderBottomWidth: 1,
    borderColor: '#E7200D',
    width: 200,
    height: 40,
  },
  upperCasestyle:{
    flexDirection:'row', 
    justifyContent:'center',
    alignItems:'center',
    color:"#FA6148"
  }
  
});

export default Home;